internal/users/locking/locking.go

```diff
@@ -37,74 +35,120 @@ var (
 	ErrLockTimeout = fmt.Errorf("%w: timeout", ErrLock)
 )
 
-func writeLockInternal() error {
-	done := make(chan error)
-	writeLockImpl := writeLockImpl
+type UserDBLock struct {
+	mu   sync.Mutex
+	cond *sync.Cond
+	held bool
+}
 
-	go func() {
-		done <- writeLockImpl()
-	}()
+// NewUserDBLock creates a new UserDBLock.
+func NewUserDBLock() *UserDBLock {
+	lock := &UserDBLock{}
+	lock.cond = sync.NewCond(&lock.mu)
+	return lock
+}
 
-	select {
-	// lckpwdf when called from cgo doesn't behave exactly the same, likely
-	// because alarms are handled by go runtime, so do it manually here by
-	// failing if "lock not obtained within 15 seconds" as per lckpwdf.3.
-	// Keep this in sync with what lckpwdf does, adding an extra second.
-	case <-time.After(maxWait):
-		return ErrLockTimeout
-	case err := <-done:
-		return err
+// Lock acquires the user database lock.
+// Returns an error if already held.
+func (l *UserDBLock) Lock() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	for l.held {
+		l.cond.Wait()
 	}
-}
 
-// WriteRecLock locks the system's user database for writing.
-// While the lock is held, all other processes trying to lock the database
-// will block until the lock is released (or a timeout of 15 seconds is reached).
-// Note that this implies that if some other process owns the lock when
-// this function is called, it will block until the other process releases the
-// lock.
-//
-// This function is recursive, it can be called multiple times without
-// deadlocking even by different goroutines - the system user database is locked
-// only once, when the reference count is 0, else it just increments the
-// reference count.
-//
-// [WriteRecUnlock] must be called the same number of times as [WriteRecLock] to
-// release the lock.
-func WriteRecLock() error {
-	writeLocksCountMu.Lock()
-	defer writeLocksCountMu.Unlock()
-
-	if writeLocksCount == 0 {
-		if err := writeLockInternal(); err != nil {
-			return err
-		}
+	if err := writeLockInternal(); err != nil {
+		return err
 	}
 
-	writeLocksCount++
+	l.held = true
+
 	return nil
 }
 
-// WriteRecUnlock decreases the reference count of the lock acquired by.
-// [WriteRecLock]. If the reference count reaches 0, it releases the lock
-// on the system's user database.
-func WriteRecUnlock() error {
-	writeLocksCountMu.Lock()
-	defer writeLocksCountMu.Unlock()
+// TryLock attempts to acquire the user database lock without blocking.
+func (l *UserDBLock) TryLock() error {
```

Comment:
I don't like try locking mechanisms, as they are prone to races, since you have potential false negative results (e.g. the if two functions call it concurrently, we end up calling `writeLockInternal` from both, one will fail (while maybe it was not a problem), but it this is no difference from just the old behavior of `userslocking.WriteLock()`.

Thus I really prefer having a ref-counted system if we want to have a no-blocking path, or not having it at all